rectools/metrics/auc.py

```diff
@@ -0,0 +1,524 @@
+    Container with meta data got from `_AUCMetric.fit` method.
+
+    Parameters
+    ----------
+    outer_merged_enriched : pd.DataFrame
+        Recommendations outer merged with test interactions. Table has Columns.User, Columns.Item,
+        Columns.Rank. Precomputed columns include "__test_positive", "__tp", "__fp", "__fp_cumsum",
+        "__test_pos_cumcum". All ranks for all users are present with no skipping.
+        Null ranks are specified for test interactions that were not predicted in recommendations.
+    n_pos : pd.Series
+        Number of positive items for each user in test insteractions.
+    n_fp_insufficient : pd.Series
+        Number of false positive items for each user in `outer_merged_enriched` that had at least
+        one false negative. This users will be checked for insufficient cases processing.
+
+    """
+
+    outer_merged_enriched: pd.DataFrame = attr.ib()
+    n_pos: pd.Series = attr.ib()
+    n_fp_insufficient: pd.Series = attr.ib()
+
+
+@define
+class _AUCMetric(MetricAtK):
+    """
+    ROC AUC based metric base class.
+
+    Warning: This class should not be used directly.
+    Use derived classes instead.
+
+    Parameters
+    ----------
+    k : int
+        Number of items at the top of recommendations list that will be used to calculate metric.
+    insufficient_handling : {"ignore", "raise", "exclude"}, default `"ignore"`
+        Method of handling users with insufficient recommendations for metric calculation.
+        ROC AUC based metrics with `k` parameter often need more then `k` recommendations
+        for each user. This happens because this metrics calculate ROC AUC
+        score for specific number of user false positives and ranked test positives that is derived
+        from provided `k` parameter but is not equal to it.
+        The following methods are available:
+        - `ignore` - don't check for insufficient recommendations lists, handle all of insufficient
+        cases as if algorithms are not able to retrieve users unpredicted test positives on any k
+        level. This will understate the metric value;
+        - `exclude` - exclude all users with insufficient recommendations lists from metrics
+        computation;
+        - `raise` - raise error if there are any users with insufficient recommendations lists. Use
+        this option very carefully because some of the algorithms are unable to provide full required
+        lists because of their inference logic. So can get errors even if you requested enough
+        recommendations in `recommend` method. For example, ItemKNN generates recommendations only
+        until the model has non-zero scores for the item in item-item similarity matrix. So with
+        small `K` for neighbours in ItemKNN and big `K` for `recommend` and AUC based metric you
+        will still get an error when `insufficient_handling` is set to `raise`.
+    """
+
+    insufficient_handling: tp.Optional[str] = field(default="ignore")
```

Comment:
Why is it `Optional`?  It will fail with `None`, right?

Btw, maybe make it  `Literal` since we rejected py3.7 support?

Reply:
Fixed optional. Discussed that Literal is not a must